### PR TITLE
Laboratory work #4, Elena Ermak - 22FPL1

### PR DESCRIPTION
**Pull request by ermakelena**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +30,20 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(text, str) or not text:
+            raise ValueError("Incorrect input")
+
+        tokens = []
+        for token in text.lower().split():
+            if token[-1] in "?!.":
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:38]: `"?!."` to a variable

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -220,6 +318,34 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
                 or if methods used return None,
                 or if nothing was generated.
         """
+        if not (isinstance(generated_text, str) and generated_text):
+            raise ValueError("Incorrect input data")
+
+        encoded_generated_text = self._word_processor.encode(generated_text)
+        if not encoded_generated_text:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:326]: message

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -259,6 +400,7 @@ def __init__(self, json_path: str) -> None:
             json_path (str): Local path to assets file
         """
 
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:403]: remove

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -274,6 +416,7 @@ def _load_from_json(self) -> dict[tuple[str, int], str]:  # type: ignore
                 or if inappropriate type loaded data.
         """
 
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:419]: remove

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -298,6 +441,7 @@ def assess_exam(self, answers: dict[str, str]) -> float:  # type: ignore
         """
 
 
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:444]: remove

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (ermakelena) [reply lab_4_fill_words_by_ngrams/main.py:40]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 2 above)
Comment (ermakelena) [reply lab_4_fill_words_by_ngrams/main.py:326]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (ermakelena) [reply lab_4_fill_words_by_ngrams/main.py:444]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (ermakelena) [reply lab_4_fill_words_by_ngrams/main.py:419]: done

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (ermakelena) [reply lab_4_fill_words_by_ngrams/main.py:403]: done

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**